include/async/then.hpp

```diff
@@ -22,47 +22,105 @@
 namespace async {
 namespace _then {
 
+namespace detail {
+template <typename... Ts> struct args {
+    using tuple_t = stdx::tuple<Ts...>;
+
+    template <typename F, std::size_t Offset, std::size_t... Is>
+    constexpr static auto try_invoke_impl(std::index_sequence<Is...>)
+        -> std::invoke_result_t<
+            F, decltype(std::declval<tuple_t>()[stdx::index<Offset + Is>])...>;
+
+    template <typename F, typename Offset, typename N>
+    using try_invoke = decltype(try_invoke_impl<F, Offset::value>(
+        std::make_index_sequence<N::value>{}));
+
+    template <typename F, typename Offset, typename N>
+    using has_arg_count = boost::mp11::mp_valid<try_invoke, F, Offset, N>;
+
+    template <typename F, typename Offset, typename N> struct arity;
+
+    template <typename F, typename Offset, typename N>
+    using arity_t = typename arity<F, Offset, N>::type;
+
+    template <typename F, typename Offset, typename N> struct arity {
+        using type = boost::mp11::mp_eval_if<
+            has_arg_count<F, Offset, N>, N, arity_t, F, Offset,
+            std::integral_constant<std::size_t, N::value + 1u>>;
+    };
+
+    template <typename... Fs>
+    using arities_t =
```

Comment:
**praise**: this is really beautiful!

Reply:
Conceptually, I agree. It's some syntax though!